fortimanager_connector.py

```diff
@@ -0,0 +1,462 @@
+                }
+        if level == 'ADOM':
+            adom = param.get('adom')
+            if not adom:
+                adom = 'root'
+            endpoint = ADOM_FIREWALL_ENDPOINT.format(adom=adom, pkg=pkg)
+        elif level == 'Global':
+            adom = 'global'
+            policy_type = param['policy_type']
+            endpoint = GLOBAL_FIREWALL_ENDPOINT.format(pkg=pkg, policy_type=policy_type)
+
+        try:
+            fmg_instance = self._login(action_result)
+            fmg_instance.lock_adom(adom)
+            response_code, response_data = fmg_instance.add(endpoint, **data)
+            fmg_instance.commit_changes(adom)
+            fmg_instance.unlock_adom(adom)
+            fmg_instance.logout()
+        except Exception as e:
+            error_msg = self._get_error_msg_from_exception(e)
+            self.save_progress("Create Firewall Policy action failed")
+            self.debug_print("Create Firewall Policy action failed: {}".format(error_msg))
+            return action_result.set_status(phantom.APP_ERROR, None)
+
+        if response_code == 0:
+            action_result.add_data(response_data)
+            summary = action_result.update_summary({})
+            summary['status'] = 'Successfully added firewall policy'
+            return action_result.set_status(phantom.APP_SUCCESS)
+        else:
+            self.save_progress("Failed.")
+            error_msg = response_data['status']['message']
+            return action_result.set_status(phantom.APP_ERROR, "Failed to create firewall policy. Reason: {}".format(error_msg))
+
+    def _handle_list_firewall_policies(self, param):
+
+        self.save_progress("In action handler for: {0}".format(self.get_action_identifier()))
+
+        action_result = self.add_action_result(ActionResult(dict(param)))
+        fmg_instance = None
+        level = param["level"]
+        pkg = param.get('package')
+        package_path = param.get('package_path')
+        policy_name = param.get('policy_name')
+        if pkg and package_path:
+            pkg += '/' + package_path
+        if level == 'ADOM':
+            adom = param.get('adom')
+            if not adom:
+                adom = 'root'
+            endpoint = LIST_ADOM_FIREWALL_POLICY.format(adom=adom, pkg=pkg)
+        elif level == 'Global':
+            endpoint = LIST_GLOBAL_FIREWALL_POLICY.format(pkg=pkg, policy_type=param.get('policy_type'))
+
+        try:
+            fmg_instance = self._login(action_result)
+            if policy_name:
+                data = {
+                    'filter': [
+                        [
+                            "name", "==", "{}".format(policy_name)
+                        ]
+                    ]}
+                response_code, firewall_policies = fmg_instance.get(endpoint, **data)
+            else:
+                response_code, firewall_policies = fmg_instance.get(endpoint)
+            fmg_instance.logout()
```

Comment:
would it be best to call `fmg_instance.logout()` outside of this `try` block in case the `fmg_instance.get()` call fails/raises an exception for some reason (so there aren't a bunch of open sessions if the action fails)?

Reply:
Good call